src/pages/poe/characters/index.tsx

```diff
@@ -104,153 +84,202 @@ export default function Characters({
-    onCompleted(data) {
-      setSearchResponse({
-        ...searchResponse,
-        ...data.characterSnapshotsSearch,
-      });
-      setAggregationSearchResponse({
-        ...aggregationSearchResponse,
-        ...data.characterSnapshotsSearchAggregations,
-      });
-    },
-  });
+
+  const { refetch: reftechGeneralSearch } = useQuery(
+    generalSearch, 
+    {
+      skip: true,
+      variables: { search: { ...search, ...{ league: league } } },
+      onCompleted(data) {
+        setCharacters({
+          ...characters,
+          ...data.characterSnapshotsSearch,
+        });
+        setAggregations({
+          ...aggregations,
+          ...data.characterSnapshotsSearchAggregations,
+        });
+      },
+    });
 
   useEffect(() => {
     reftechGeneralSearch();
   }, [search, reftechGeneralSearch, league]);
 
-  const updateIncludeExclude = (entry, includedKey, excludedKey) => {
-    if (search[includedKey]?.includes(entry.key)) {
-      setSearch({
-        ...search,
-        ...{
-          [excludedKey]: [...search[excludedKey]!, entry.key],
-          [includedKey]: search[includedKey]!.filter((e) => e !== entry.key),
-        },
-      });
-    } else if (search[excludedKey]?.includes(entry.key)) {
-      setSearch({
-        ...search,
-        ...{
-          [excludedKey]: search[excludedKey]!.filter((e) => e !== entry.key),
-        },
-      });
-    } else {
-      setSearch({
-        ...search,
-        ...{
-          [includedKey]: [...search[includedKey]!, entry.key],
-        },
-      });
-    }
-  };
-
-  if (!searchResponse) {
+  if (!characters) {
     return <>Loading...</>;
   }
 
+  /*
+   * Define callbacks as functions instead of consts
+   * to skip revaluation on every component view update.
+   */
+
+  function updateIncludeExclude (
+    search: CharacterSnapshotSearch, 
+    setSearch:Dispatch<any>, 
+    entry: {key: string, value: any}, 
+    includedKey: string, 
+    excludedKey: string): void {
+  
+      if (search[includedKey]?.includes(entry.key)) {
+        setSearch({
+          ...search,
+          ...{
+            [excludedKey]: [...search[excludedKey]!, entry.key],
+            [includedKey]: search[includedKey]!.filter((e: string) => e !== entry.key),
+          },
+        });
+      } else if (search[excludedKey]?.includes(entry.key)) {
+        setSearch({
+          ...search,
+          ...{
+            [excludedKey]: search[excludedKey]!.filter((e: string) => e !== entry.key),
+          },
+        });
+      } else {
+        setSearch({
+          ...search,
+          ...{
+            [includedKey]: [...search[includedKey]!, entry.key],
+          },
+        });
+      }
+  }
+
+  function onSkillChange(mainSkill: {key: string, value: any}) {
+    updateIncludeExclude( search, setSearch,
+      mainSkill,
+      "includedMainSkills",
+      "excludedMainSkills"
+    )
+  }
+
+  function onClassChange(characterClass: {key: string, value: any}) {
```

Comment:
It seems like there should be a way to do this without having to declare a method for each of these.